migrations/__init__.py

```diff
@@ -44,8 +44,14 @@ def migrate(db: DuckDBPyConnection, from_version: Optional[int] = None, to_versi
     if from_version > to_version:
         raise ValueError("Cannot migrate from a higher version to a lower version.")
 
-    for version in range(from_version, to_version):
+    # dont migrate if already at the target version
+    if from_version == to_version:
+        return from_version, to_version
+
+    print(f"Migrating from version {from_version} to {to_version}...")
+    for version in range(from_version, to_version + 1):
         if version in MIGRATIONS:
```

Comment:
That condition is wrong in the original code, we should be checking if `version + 1` is in `MIGRATIONS` if that's the migration we are trying to apply later. Although that should not cause any issues if `migrate()` is called without providing `from_version` and `to_version`, because `to_version` obviously exists then.

Fixing this condition also removes the need for the `if from_version == to_version` block, because `range(from_version, to_version)` would yield nothing if `from_version` is equal to `to_version`.

Reply:
Oh, I see! Good call. That makes sense! Apologies on the confusion there. 

For me (using the main branch), I was getting this error: 

```text
[2023-05-20 10:48:19:386510] - run - [INFO] Running migrations...
Traceback (most recent call last):
  File "/home/justin/.pyenv/versions/3.10.5/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/justin/.pyenv/versions/3.10.5/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/justin/Documents/opensource/matrix-gptbot/gptbot.py", line 33, in <module>
    asyncio.run(bot.run())
  File "/home/justin/.pyenv/versions/3.10.5/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/home/justin/.pyenv/versions/3.10.5/lib/python3.10/asyncio/base_events.py", line 646, in run_until_complete
    return future.result()
  File "/home/justin/Documents/opensource/matrix-gptbot/classes/bot.py", line 591, in run
    before, after = migrate(self.database)
  File "/home/justin/Documents/opensource/matrix-gptbot/migrations/__init__.py", line 49, in migrate
    MIGRATIONS[version + 1](db)
  File "/home/justin/Documents/opensource/matrix-gptbot/migrations/migration_2.py", line 133, in migration
    cursor.execute(
duckdb.CatalogException: Catalog Error: Table with name migrations does not exist!
Did you mean "accounts"?
```

After what you mentioned, I tried this:

```python
    for version in range(from_version, to_version):
        if version + 1 in MIGRATIONS:
            MIGRATIONS[version + 1](db)

    return from_version, to_version
```

Seems to have done the trick.

```text
[2023-05-20 10:49:41:247924] - from_config - [INFO] Loading logo from /home/justin/Documents/opensource/matrix-gptbot/assets/logo.png
[2023-05-20 10:49:41:524727] - run - [INFO] Running migrations...
[2023-05-20 10:49:41:588611] - run - [INFO] Migrated from version 0 to 8.
```

Reply:
Ah, yes, that makes sense. With the logic as it was previously implemented, the first value of `version` would have been 0 - it would then check if `MIGRATION[0]` exists, and because it doesn't, `MIGRATION[0 + 1]`, the first migration, in which the `migrations` table is created, does not get applied.